Fixed the Dune 2000 harvester sand out animation

Mailaender opened the pull request:
using a new blend mode from https://github.com/OpenRA/OpenRA/pull/7811 to make it look more like the original.

Mailaender committed:
add alpha channel to every possible blend mode

pchote commented:
This looks pretty broken (nothing like the original) on my machine:

![screen shot 2015-04-19 at 18 26 27](https://cloud.githubusercontent.com/assets/167819/7218420/52fd2bfe-e6c2-11e4-8f20-c65c5dbae1bf.png)

penev92 commented:
Just for reference, in Dune 2000 it looks like this:
![](http://i.imgur.com/Ryp6FdI.png)        into        ![](http://i.imgur.com/rNr7ioW.png)

And on bleed it's like this:
![](http://i.imgur.com/leojETf.png)

and no zoom:
![](http://i.imgur.com/IcuEG7o.png)

Mailaender commented:
I was trying to get rid of the hard edges

![image](https://cloud.githubusercontent.com/assets/756669/7218808/36591328-e681-11e4-92a6-0be74555f92e.png)

Mailaender committed:
fix the Multiply blend mode

Mailaender commented:
Cleaned up the per-sequence transparency for more flexibility and updated with the help of @tomsons26 and some quick research at http://nial.me/2012/06/porting-photoshop-blendmodes-to-cocos2d/

Mailaender committed:
fix d2k harvester sand overlay

Mailaender commented:
Rebased.

pchote commented on OpenRA.Game/Graphics/LineRenderer.cs:
```diff
@@ -52,7 +52,7 @@ public void Flush()
 			if (nv > 0)
 			{
-				renderer.Device.SetBlendMode(BlendMode.Alpha);
+				renderer.Device.SetBlendMode(BlendMode.Alpha, 1f);
```
This is ugly.  Set the default value in the interface instead.

Mailaender replied:
Done.

pchote commented on OpenRA.Game/Graphics/SheetBuilder.cs:
```diff
@@ -66,7 +66,7 @@ public Sprite Add(byte[] src, Size size, float2 spriteOffset)
 			// Don't bother allocating empty sprites
 			if (size.Width == 0 || size.Height == 0)
-				return new Sprite(current, Rectangle.Empty, spriteOffset, channel, BlendMode.Alpha);
+				return new Sprite(current, Rectangle.Empty, spriteOffset, channel, BlendMode.Alpha, 1f);
```
alpha = 1 is already the default parameter value, so you don't need to specify it again here.

pchote commented on OpenRA.Game/Graphics/SheetBuilder.cs:
```diff
@@ -129,7 +129,7 @@ public Sprite Allocate(Size imageSize, float2 spriteOffset)
 			}
 
-			var rect = new Sprite(current, new Rectangle(p, imageSize), spriteOffset, channel, BlendMode.Alpha);
+			var rect = new Sprite(current, new Rectangle(p, imageSize), spriteOffset, channel, BlendMode.Alpha, 1f);
```
or here.

pchote commented on OpenRA.Game/Graphics/SpriteRenderer.cs:
```diff
@@ -58,6 +59,7 @@ void SetRenderStateForSprite(Sprite s)
 				Flush();
 
 			currentBlend = s.BlendMode;
+			alpha = s.Alpha;
```
You need to Flush() if the alpha changes.

pchote commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -119,12 +119,13 @@ public DefaultSpriteSequence(ModData modData, TileSet tileSet, SpriteCache cache
 
 				var offset = LoadField<float2>(d, "Offset", float2.Zero);
 				var blendMode = LoadField<BlendMode>(d, "BlendMode", BlendMode.Alpha);
+				var alpha = LoadField<float>(d, "Alpha", 1f);
```
The blend mode is called "Translucency" so it may be clearer to call this "Translucency" too.

Mailaender replied:
This is not tied to any blend mode. OpenTK, OpenGL and PhotoShop seem to call it alpha (channel).

pchote commented on OpenRA.Renderer.Sdl2/Sdl2GraphicsDevice.cs:
```diff
@@ -246,26 +246,11 @@ public void SetBlendMode(BlendMode mode)
 					GL.BlendFunc(BlendingFactorSrc.OneMinusDstColor, BlendingFactorDest.One);
 					break;
-				case BlendMode.Translucency25:
+				case BlendMode.Translucency:
```
Having a dedicated Translucency blend mode but then allowing any other mode to set its own Alpha (translucency) value smells a bit like spaghetti.

Mailaender replied:
I would otherwise have to create a redundant Multiply50 blend mode.

pchote commented:
I'm not convinced about these alpha changes, but will keep an eye on how this turns out.

obrakmann commented:
FWIW, I feel it makes more sense to have single a configurable translucency mode than having multiple hard-coded ones. But I don't know too much about any of the graphics stuff.

Anyway, :+1: from me.

abcdefg30 commented:
Looks also good to me.
:+1: / :white_check_mark:

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/81db847a35159b6eb795ac80661908107b68527d...73378215a3938eb74b63c87bbd7f289681e9b5ac)
